user/xargs.c

```diff
@@ -0,0 +1,161 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+#define MAX_LINE 512
+/**
+ * this program reads line by line from stdin, and executes its arguments with the arguments from stdin appended.
+*/
+#define min(a,b) ((a) < (b) ? (a) : (b))
+
+
+char WHITESPACE_CHARS[] = { ' ', '\t', '\n' };
+
+int is_whitespace(char c)
+{
+  for (uint indx = 0; indx < sizeof(WHITESPACE_CHARS); indx++)
+  {
+    if (c == WHITESPACE_CHARS[indx]) return 1;
+  }
+  return 0;
+}
+
+int not_whitespace(char c)
+{
+  for (uint indx = 0; indx < sizeof(WHITESPACE_CHARS); indx++)
+  {
+    if (c == WHITESPACE_CHARS[indx]) return 0;
+  }
+  return 1;
+}
+
+int skip_whitespace(char *text, char **result)
+{
+  char *start = text;
+  while ((*text != 0) && is_whitespace(*text))
+    text++;
+  *result = text;
+  return start - text;
+}
+
+int skip_arg(char *text, char **result)
+{
+  char *start = text;
+  while ((*text != 0) && not_whitespace(*text))
+    text++;
+  *result = text;
+  return text - start;
+}
+
+int count_args(char *line)
+{
+  int arg_count = 0;
+  skip_whitespace(line, &line);
+  while (*line != 0)
+  {
+    arg_count++;
+    skip_arg(line, &line);
+    skip_whitespace(line, &line);
+  }
+  return arg_count;
+}
+
+int parse_args(char *line, int old_argc, char **old_argv, char ***new_argv_result)
+{
+  int line_arg_amnt = count_args(line);
+  int new_argc = old_argc + line_arg_amnt;
+  char **new_argv = malloc((new_argc + 1) * sizeof(char *));
+  for (int indx = 0; indx < old_argc; indx++)
+  {
+    new_argv[indx] = malloc(strlen(old_argv[indx]) * sizeof(char));
+    strcpy(new_argv[indx], old_argv[indx]);
+  }
+  char *current_arg = line;
+  skip_whitespace(current_arg, &current_arg);
+  for (int indx = old_argc; indx < new_argc; indx++)
+  {
+    char *arg_start = current_arg;
+    int len = skip_arg(current_arg, &current_arg);
+    skip_whitespace(current_arg, &current_arg);
+    new_argv[indx] = malloc((len + 1) * sizeof(char));
+    strncpy(new_argv[indx], arg_start, len);
+    new_argv[indx][len] = '\0';
+  }
+  new_argv[new_argc] = 0;
+  *new_argv_result = new_argv;
+  return new_argc;
+}
+
+
+void free_argv(char **argv, int argc)
+{
+  for (int indx = 0; indx < argc; indx++)
+  {
+    free(argv[indx]);
+  }
+  free(argv);
+}
+
+void *realloc(void *mem, int old_size, int new_size)
+{
+  void *new_mem = malloc(new_size);
+  memmove(new_mem, mem, min(old_size, new_size));
+  free(mem);
+  return new_mem;
+}
+
+int getline(char **line, int fd)
+{
+  uint capacity = 10;
```

Comment:
what about MAXARG ?

Reply:
i mean then the exec would fail - why should i treat it differently